dm/syncer/validate_worker.go

```diff
@@ -0,0 +1,380 @@
+	tidbmysql "github.com/pingcap/tidb/parser/mysql"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/tiflow/dm/dm/config"
+	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/dm/pkg/terror"
+	"github.com/pingcap/tiflow/dm/syncer/dbconn"
+)
+
+const (
+	workerChannelSize = 1000
+
+	maxBatchSize = 100 // todo: choose a proper value, or configurable?
+)
+
+type validateFailedType int
+
+const (
+	hasDeletedRow validateFailedType = iota
+	rowNotExist
+	rowDifferent
+)
+
+type validateFailedRow struct {
+	tp      validateFailedType
+	dstData []*sql.NullString
+}
+
+type validateWorker struct {
+	cfg               config.ValidatorConfig
+	ctx               context.Context
+	interval          time.Duration
+	validator         *DataValidator
+	L                 log.Logger
+	conn              *dbconn.DBConn
+	rowChangeCh       chan *rowChange
+	pendingChangesMap map[string]*tableChange
+	pendingRowCount   atomic.Int64
+	batchSize         int
+	sync.Mutex
+}
+
+func newValidateWorker(v *DataValidator, id int) *validateWorker {
+	return &validateWorker{
+		cfg:               v.cfg.ValidatorCfg,
+		ctx:               v.ctx,
+		interval:          v.validateInterval,
+		validator:         v,
+		L:                 v.L,
+		conn:              v.toDBConns[id],
+		rowChangeCh:       make(chan *rowChange, workerChannelSize),
+		pendingChangesMap: make(map[string]*tableChange),
+		batchSize:         maxBatchSize,
+	}
+}
+
+func (vw *validateWorker) run() {
+outer:
+	for {
+		select {
+		case change, ok := <-vw.rowChangeCh:
+			if !ok {
+				break outer
+			}
+			// todo: limit number of pending rows.
+			// todo: trigger validation when pending rows count reaches a threshold, but we don't want trigger it too often
+			// todo: since it may not reduce the number of pending row changes as the downstream may have changed again,
+			// todo: we need to catchup with the progress of syncer first, so row changes which are changed again can be merged.
+			vw.updateRowChange(change)
+		case <-time.After(vw.interval):
+			err := vw.validateTableChange()
+			if err != nil {
+				// todo: better error handling
+				vw.validator.errChan <- terror.Annotate(err, "failed to validate table change")
+			}
+			// todo: if a row failed too many times, add it to failed changes
+		}
+	}
+}
+
+func (vw *validateWorker) updateRowChange(row *rowChange) {
+	// cluster using target table
+	fullTableName := row.table.Target.String()
+	change := vw.pendingChangesMap[fullTableName]
+	if change == nil {
+		// no change of this table
+		change = &tableChange{
+			table: row.table,
+			rows:  make(map[string]*rowChange),
+		}
+		vw.pendingChangesMap[fullTableName] = change
+	}
+	if val, ok := change.rows[row.key]; ok {
+		val.data = row.data
+		val.tp = row.tp
+		val.lastMeetTS = row.lastMeetTS
+		val.failedCnt = 0 // clear failed count
+	} else {
+		change.rows[row.key] = row
+		vw.pendingRowCount.Inc()
```

Comment:
why we need this `pendingRowCount`  i don'e see any usage of this var

Reply:
need to print internal status regularly or send metrics to prometheus later, not used now.